core/node/provider.go

```diff
@@ -2,50 +2,129 @@ package node
 
-	"github.com/libp2p/go-libp2p-routing"
 	"go.uber.org/fx"
 
+	"github.com/ipfs/go-ipfs-config"
 	"github.com/ipfs/go-ipfs/core/node/helpers"
 	"github.com/ipfs/go-ipfs/provider"
+	q "github.com/ipfs/go-ipfs/provider/queue"
+	"github.com/ipfs/go-ipfs/provider/simple"
 	"github.com/ipfs/go-ipfs/repo"
-	"github.com/ipfs/go-ipfs/reprovide"
+	"github.com/libp2p/go-libp2p-routing"
 )
 
 const kReprovideFrequency = time.Hour * 12
 
+// SIMPLE
+
 // ProviderQueue creates new datastore backed provider queue
-func ProviderQueue(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo) (*provider.Queue, error) {
-	return provider.NewQueue(helpers.LifecycleCtx(mctx, lc), "provider-v1", repo.Datastore())
+func ProviderQueue(mctx helpers.MetricsCtx, lc fx.Lifecycle, repo repo.Repo) (*q.Queue, error) {
+	return q.NewQueue(helpers.LifecycleCtx(mctx, lc), "provider-v1", repo.Datastore())
 }
 
-// ProviderCtor creates new record provider
-func ProviderCtor(mctx helpers.MetricsCtx, lc fx.Lifecycle, queue *provider.Queue, rt routing.IpfsRouting) provider.Provider {
-	p := provider.NewProvider(helpers.LifecycleCtx(mctx, lc), queue, rt)
+// SimpleProviderCtor creates new record provider
+func SimpleProviderCtor(mctx helpers.MetricsCtx, lc fx.Lifecycle, queue *q.Queue, rt routing.IpfsRouting) provider.Provider {
+	return simple.NewProvider(helpers.LifecycleCtx(mctx, lc), queue, rt)
+}
+
+// SimpleReproviderCtor creates new reprovider
+func SimpleReproviderCtor(reproviderInterval time.Duration) func(helpers.MetricsCtx, fx.Lifecycle, routing.IpfsRouting, simple.KeyChanFunc) (provider.Reprovider, error) {
+	return func(mctx helpers.MetricsCtx, lc fx.Lifecycle, rt routing.IpfsRouting, keyProvider simple.KeyChanFunc) (provider.Reprovider, error) {
+		return simple.NewReprovider(helpers.LifecycleCtx(mctx, lc), reproviderInterval, rt, keyProvider), nil
+	}
+}
 
+// SimpleProviderSysCtor creates new provider system
+func SimpleProviderSysCtor(lc fx.Lifecycle, p provider.Provider, r provider.Reprovider) provider.System {
+	sys := provider.NewSystem(p, r)
 	lc.Append(fx.Hook{
 		OnStart: func(ctx context.Context) error {
-			p.Run()
+			sys.Run()
 			return nil
 		},
 		OnStop: func(ctx context.Context) error {
-			return p.Close()
+			return sys.Close()
 		},
 	})
+	return sys
+}
+
+// SimpleOfflineProviderSysCtor creates a new offline provider system
+func SimpleOfflineProviderSysCtor(p provider.Provider, r provider.Reprovider) provider.System {
+	return provider.NewSystem(p, r)
+}
+
+// STRATEGIC
+
+// StrategicProviderSysCtor creates new provider system
+func StrategicProviderSysCtor() provider.System {
+	return provider.NewOfflineProvider()
+}
+
+// StrategicOfflineProviderSysCtor creates a new offline provider system
+func StrategicOfflineProviderSysCtor() provider.System {
+	return provider.NewOfflineProvider()
+}
+
+// ONLINE/OFFLINE
+
+// OnlineProviders groups units managing provider routing records online
+func OnlineProviders(cfg *config.Config) fx.Option {
```

Comment:
@magik6k I also want to point out that I moved the `Providers` function (used for DI) to the provider node. I did this because there's a bit of logic I wanted to keep near the provider code (like taking into account the experimental flag).

Reply:
Anything that touches config should generally live in `groups.go`. This might seem a bit weird right now, but hopefully will make much more sense after the constructor refactors are finished